packages/block-library/src/navigation/index.php

```diff
@@ -132,6 +132,92 @@ function block_core_navigation_render_submenu_icon() {
 	return '<svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 12 12" fill="none" role="img" aria-hidden="true" focusable="false"><path d="M1.50002 4L6.00002 8L10.5 4" stroke-width="1.5"></path></svg>';
 }
 
+
+/**
+ * Finds the first non-empty `wp_navigation` Post.
+ *
+ * @return WP_Post|null the first non-empty Navigation or null.
+ */
+function block_core_navigation_get_non_empty_navigation() {
+	// Order and orderby args set to mirror those in `wp_get_nav_menus`
+	// see:
+	// - https://github.com/WordPress/wordpress-develop/blob/ba943e113d3b31b121f77a2d30aebe14b047c69d/src/wp-includes/nav-menu.php#L613-L619.
+	// - https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters.
+	$navigation_posts = get_posts(
```

Comment:
By default `get_posts` won't return every single post. (I think the default is 5?) So the logic error in this function is that, if there are `n` navigation posts, and `n+1` is the non-empty one, this function will incorrectly return `null`.

We could fix this by setting `posts_per_page` to `-1` to return all posts, but this isn't a good idea as there could be more posts than the server has RAM.

We need to encode the _get the first non-empty post_ logic into the query itself. Unfortunately `WP_Query` doesn't have an argument which equates to `post_content != ''`. But that's ok, we know that in this case a non-empty navigation post will contain block markup so we could instead search for `<!--`.

Something like this:

```php
function block_core_navigation_get_non_empty_navigation() {
	$navigation_posts = get_posts(
			array(
 				'post_type'      => 'wp_navigation',
	 			'order'          => 'ASC',
	 			'orderby'        => 'name',
				'posts_per_page' => 1,
				's'              => '<!--',
	 		)
	);
	return count( $navigation_posts ) ? $navigation_posts[0] : null;
}
```

https://developer.wordpress.org/reference/classes/wp_query/parse_query/

Reply:
Great insight. I'll start work on this now. Thank you @noisysocks!

Reply:
This works well as far as I can see.

The original implementation was mirroring what is in `wp_nav_menu()`. The thought process was that it is highly unlikely that you will have 100s of `wp_navigation` posts so the in memory processing would probably never become an issue....probably.

Overall, so long as the `LIKE` in the query doesn't degrade the query performance we should be ok with your revised solution. It's certainly more elegant and avoids the looping that I had in the original implementation.

During my testing this seemed to work well. I appreciate your insight here 🙇‍♂️

Reply:
Let's not use get_posts where possible. `get_posts` is not cachable and does not run through filters.